src/middleware/redux.ts

```diff
@@ -1,29 +1,20 @@
 import { StateCreator, StoreMutatorIdentifier } from '../vanilla'
 import { NamedSet } from './devtools'
 
-type Write<T extends object, U extends object> = Omit<T, keyof U> & U
-type Cast<T, U> = T extends U ? T : U
+type Write<T, U> = Omit<T, keyof U> & U
 
-type Action = {
-  type: unknown
-}
-
-type ReduxState<A extends Action> = {
+type ReduxState<A> = {
   dispatch: StoreRedux<A>['dispatch']
 }
 
-type StoreRedux<A extends Action> = {
+type StoreRedux<A> = {
   dispatch: (a: A) => A
   dispatchFromDevtools: true
 }
 
-type WithRedux<S, A> = Write<Cast<S, object>, StoreRedux<Cast<A, Action>>>
+type WithRedux<S, A> = Write<S, StoreRedux<A>>
 
-type Redux = <
-  T extends object,
-  A extends Action,
-  Cms extends [StoreMutatorIdentifier, unknown][] = []
->(
+type Redux = <T, A, Cms extends [StoreMutatorIdentifier, unknown][] = []>(
```

Comment:
@dai-shi Is removing the constraint in `A` intentional? This would allow non-redux actions (and non-redux reducers), which might not work well with `devtools`. I don't think it's a big deal to remove the constraint, but I'd prefer to keep it.

Reply:
Yes, this is intentional. (to remove `Cast`, which is more important for me.)
For 99% use cases, there should be no issues. btw, we have a type guard and work around to force `{ type: unknown }`.

Reply:
> to remove `Cast`, which is more important for me.

Instead of `Cast<A, Action>` we can write `A & Action` too. We just have to satisfy the type-checker. Both `Cast<A, Action>` and `A & Action` are subtype of `Action` (which is what the type-checker wants), and when `A` is subtype of `Action` (which it'll always be) both resolve to just `A`.

> btw, we have a type guard and work around to force `{ type: unknown }`.

That makes it worse, right now we're sending `isObjectWithTypeProperty(x) ? x : { type: x }` that means if someone has an action `{ kind: "INCREMENT" }` (which is now allowed) what will get sent is `{ type: { kind: "INCREMENT" } }`.


> For 99% use cases, there should be no issues.

That's true. It's not a huge problem, it's just a matter of correctness. And I'm fine eitherway.

Reply:
> if someone has an action `{ kind: "INCREMENT", by: 10 }` (which is now allowed)

I implicitly and intentionally make it loose. Redux Devtools allows `{ type: { kind: "INCREMENT", by: 10 } }`, right?
(So, I really wanted to make `State=unknown` and `Action=unknown`.)

Reply:
> Redux Devtools allows `{ type: { kind: "INCREMENT", by: 10 } }`, right?

Yes but the action is `{ kind: "INCREMENT", by: 10 }` and not `{ type: { kind: "INCREMENT", by: 10 } }`. So when the user jumps or dispatches an action from the redux devtools, what we'll receive from the extenstion is `{ type: { kind: "INCREMENT", by: 10 } }` which we'll dispatch and forward to the reducer but the reducer will simply won't recognise it.

The point I'm trying to make is all tooling, libraries, code around redux is written keeping in mind that the action will have a `type` property, so we can't allow violating the redux spec.

I personally see no point in violating the spec just for removing `& Action` from our code. But I'm fine if you're making that choice conciously. In that case maybe let's send `typeof x === "object" ? x : { type: x }` instead.

Reply:
> which we'll dispatch and forward to the reducer but the reducer will simply won't recognise it.

Ah, I overlooked that.

Hm, I thought "our" redux middleware doesn't need to follow the `type` property spec, but this your comment makes me change the idea. Let me see how `& Action` works.

Reply:
#1206